src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java

```diff
@@ -149,21 +157,32 @@ public void start(Map<String, String> properties) {
                 timestampColumns
             );
           }
-          partition = Collections.singletonMap(
-              JdbcSourceConnectorConstants.TABLE_NAME_KEY,
-              tableOrQuery
-          );
+          tablePartitionsToCheck = tableToPartitions.get(tableOrQuery);
           break;
         case QUERY:
           partition = Collections.singletonMap(
               JdbcSourceConnectorConstants.QUERY_NAME_KEY,
               JdbcSourceConnectorConstants.QUERY_NAME_VALUE
           );
+          tablePartitionsToCheck = new ArrayList<>();
+          tablePartitionsToCheck.add(partition);
           break;
         default:
           throw new ConnectException("Unexpected query mode: " + queryMode);
       }
-      Map<String, Object> offset = offsets == null ? null : offsets.get(partition);
+
+      //select the first matching offset
+      // This helps in handling any backward compatible offset changes
```

Comment:
We need a better comment:
```
// The partition map varies by offset protocol. Since we don't know which protocol each table's offsets
// are keyed by, we need to use the different possible partitions (which are in newest protocol version first)
// to find the actual offsets for each table.